### DISCUSSION TITLE
why is cell voltage above configured value in utils.py?

### DISCUSSION THREAD

**Discussion by arrow1800** [Louisvdw/dbus-serialbattery#246] [Q&A] [answered]
hi,

for starters; i have read the FAQ (including high voltage part) and the CCCM topic. those descriptions are more steering towards making sure you have a balanced cell pack. but what if my cells are somewhat out of balance (due to excessive ch/dc cycles)

I am trying to understand how exactly this driver works. When i configured the max and min Cell values in utils.py, i expected the driver to interfere with the charging algorithm as soon as one off the cells gets above the maximum (for example stop charging). (max is set on 3.40)

the CCCM algorithm is working. (Charge current is lowered accordingly to SOC). but why is cell voltage still above max value, is it more like a PID algorithm (with some overshoot)? or does the driver only look at the total bank voltage (16x3.4) ? i guess my question is; what is the expected/designed behavior?

![image](https://user-images.githubusercontent.com/12893401/194848331-a33d0e1e-be28-4a6b-a916-d14f383f61c5.png)

**Comment by Louisvdw** [comment, marked as answer]
The current logic does not have extra checks if once cell go over the cell max voltage. It use the max cell times the cell count to instruct the chargers the battery voltage that we want. 
There was some talks to improve the CCCM to look at individual cell overvoltage and reduce the charge current, but that feature improvement has not been coded yet.

**Reply by arrow1800** [reply]
ok, i guess that is the answer for now, to bad! that was the whole point for adding this driver in the first place. (i already have a shunt which is monitoring bank voltage levels) 

can you point me in the right direction? which method or file should i change to add this behavior? (python coder myself as well)

**Reply by arrow1800** [reply]
PS. i created an issue for this. #249
